### DISCUSSION TITLE
Code Styles Conventions: Questions and Proposal to Enforce Validation via CI

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#5] [Q&A]
Hi Christian,

A proposal to enforce code styles validation on every commit and PR, to preserve code consistency in the project; and a few thoughts about indentation-only lines found in PML sources...

In PR #1, when I set in [`.editorconfig`][.editorconfig] the code styles conventions for `.pml` files, I didn't enable the `trim_trailing_whitespace` option:

```editorconfig
## PPL Sources
##############
[*.ppl]
indent_style = space
indent_size = 4
end_of_line = unset
charset = utf-8
trim_trailing_whitespace = unset
insert_final_newline = true
```

The reason is that I noticed that all PML sources preserve indentation in "empty line" (which are then "indentation only" lines). I didn't want to change the default styles for PML sources, but I wanted to share some thoughts on this.

Although there's nothing wrong with preserving indentation in empty lines, and indeed it's common practice in many editors (mostly, Scintilla based editors), there are some disadvantages that come with it.

In my repositories I usually exploit the `.editorconfig` settings to also validate commits and PRs, using Travis CI and the [EClint] tool. The advantage is that any contribution (PR or commit) will be auto-validated and any code styles inconsistencies will show up directly on GitHub (the Travis build fails). This preserves code consistency in the repository, on the one hand, and prevents noise in commit diffs (e.g. whitespace/indentation difference), which can make tracking changes harder.

So, if a contributor doesn't have an editor/IDE capable of picking up the EditorConfig settings and, for example, has custom editor settings that would change all indentations from spaces to tabs at save time, similar commits (carrying lots of spurious noise) are intercepted — the risk is that some files containing only white space changes can end up in a PR commit (something I've seen happening quite often without these safeguards in place).

The problem with allowing indentation only lines is that it forces you to disable the `trim_trailing_whitespace` option in `.editorconfig`, which means that all trailing whitespace will go unnoticed in PR and commits. Also, in case of editors that don't support preserving indentation this way, it makes it harder to create clean commits.

E.g. my editor (Sublime Text) is set by default to remove all trailing whitespace on save, which means that whenever I edit a PML source I then have to manually remove from Git's work area the hunks that affect indentation lines only, one by one, in order to keep only the relevant changes in my PRs. Now I'm working on a custom package for PML, which will allow me to set how Sublime Text should handle empty lines between indented code line (i.e. preserve the indentation or remove it).

This empty line indentation is typical of editors using the Scintilla component, and I'm not sure if this can actually be disable in the editor's preferences (it might depend on whether individual editors exposes this setting or not).

I personally think that enforcing the `trim_trailing_whitespace` rule is always better. But, if this can't be done for whatever reason, I've already come up with a code styles validation solution via GREP, which will check for trailing spaces only in lines that contain at least one non-WS character. Here's an example of this type of script, from another repository, for a language whose native IDE preserves indentation-only lines:

- https://github.com/fantaisie-software/purebasic/blob/devel/validate.sh#L54

In that repository, we decided to abide to default settings of the official IDE, since most contributors would be using that to edit source files.

So, I wanted to ask you:

1. Whether I should set my PML syntax for Sublime Text to preserve indentation-only lines, trim them, or just let end users decide. (in other words, is there an idiomatic position on this?)
2. Would it be possible to add the `trim_trailing_whitespace` rule for PML sources in this repo (and remove all those indentation), so that end users' editors will pick it up from the `.editorconfig` file automatically, thus overriding their default settings when editing this project. (bare in mind that EditorConfig doesn't allow to make an exception for indentation-only lines, is either trim all trailing spaces or ignore them).
3. What do you think about enabling code styles validation at push/PR time, via Travis + ECLint?

Thanks.

[.editorconfig]: https://github.com/tajmone/converter/blob/8c20cce15bbe2823f1479e50e046cd2f29620f8a/.editorconfig#L40 "View source for PR commit"

[EClint]: https://www.npmjs.com/package/eclint "EClint page at NPM"

**Comment by ppl-lang** [comment]
Hi Tristano,

Thanks for your valuable comment, and for having replaced some remaining tabs by 4 spaces.

The indents in empty lines of the `.ppl` files are due to my editor's default behavior. However, whitespace is ignored in PPL, so there is no reason to keep whitespace before new lines.

I totally agree with you. Trailing whitespace should be removed automatically, because:
- As explained very well in your comment, we avoid "noise in commit diffs (e.g. whitespace/indentation difference)". And we "preserve code consistency in the repository".
- Editor support for `.editorconfig` is good. According to its website it's natively integrated in some editors, and there are plugins for others.
- Good auto-validation supports the important "Fail Fast!" principle.

So: Yes, please change the setting to `trim_trailing_whitespace = true`.

I will activate support for `.editorconfig` in my environment too (IntelliJIdea and VSCode).

Thanks

**Reply by tajmone** [reply]
OK, done (see PR #6).

> I will activate support for .editorconfig in my environment too (IntelliJIdea and VSCode).

These editors have excellent support for EditorConfig, so they should just pick-up the setting from the `.editorconfig` file and use those when editing this project, regardless of you default editor settings.

In my PR I was thinking of enabling code styles validation via Travis CI, which would allow to automatically intercept any commit or PR that violates code styles conventions. So, I've carried out some local tests, but it turns out it might not be doable.

The problem is that some PPL sources contain some lines with verbatim documentation examples, or command line output, which obviously can't abide to the strict indentation rules due formatting necessity. The files are:

- `work/src/PPL/dev/pml/converter/se_CLI_start.ppl`
- `work/src/PPL/dev/pml/converter/data/formal_node/block/with_child_nodes/monospace/se_PML_monospace_formal_node.ppl`

where the latter contains a Pascal Triangle example:

```ppl
                examples = '''[monospace title = "A Pascal Triangle"
      1
    1   1
  1   2   1
1   3   3   1
]''' ) )
.
```

Obviously it's impossible to force similar examples to meet the strict validation criteria of tools like [ECLint], since these validation tools are "dumb" and follow the rules blindly. One could disable the `indent_size` rule for these specific files in `.editorconfig`, but this would also disable it in the editor (unfortunately there are no validation tools that allow using a separate settings file from the one used by the EditorConfig plug-ins).

Since PPL is strictly tied to PML, the precedence should go to freely including PML examples in multi-line strings or block comments.

(Anyway, this is actually a common problem with all languages that have comment blocks, or with syntaxes like markdown or AsciiDoc, which might contain code blocks from other languages, where these snippets have their own code styles.)

Maybe I can come up with a scripted solution, e.g. filtering the reported errors via a manually maintained exclusion list, which would ignore EClint warnings for specific files (thus we don't need to change the `.editorconfig`). I'll let you know if I find an alternative solution, after all it would be nice to have Travis CI warning on GitHub when commits and PRs contain noisy code changes; it would allow contributors to polish their PRs before merging them (sifting through commits diffs is already a hard enough task, spurious noise just makes it unbearable).

In any case, with these `.editorconfig` settings in place now, it's enough for end users to work with an EditorConfig compliant editor to ensure clean commits/PRs and preserve code styles consistency in the project. This is the most important part — auto-validation would have been and added benefit, but it's not essential.



[ECLint]: https://www.npmjs.com/package/eclint

**Reply by pml-lang** [reply]
> enabling code style validation via Travis CI

We must indeed keep the freedom to include any text blocks in documentation strings embedded in source code.
That makes it hard to apply automatic code style validation, as you correctly pointed out.

Let's keep it "as simple as possible but not simpler".

Your contribution to `.editorconfig` is a great addition to ensure some source code consistency and avoid unnecessary complications when merging commits.

Great job. Thank you.
